src/java.base/share/native/launcher/main.c

```diff
@@ -278,15 +321,32 @@ main(int argc, char **argv)
         margv = args->elements;
     }
 
-    // Avoid unexpected process completion when checkpointing under docker container run
-    // by creating the main process waiting for children before exit.
-    if (is_checkpoint && 1 == getpid()) {
-        g_child_pid = fork();
-        if (0 < g_child_pid) {
-            // The main process should forward signals to the child.
-            setup_sighandler();
-            const int status = wait_for_children();
-            exit(status);
+    if (is_checkpoint) {
+        const int crac_min_pid_default = 128;
+        const char *env_min_pid_str = getenv("CRAC_MIN_PID");
+        const int env_min_pid = env_min_pid_str ? atoi(env_min_pid_str) : 0;
+        // TODO: should it be checked for max pid overflow?
```

Comment:
As far as I understand the max PID is not INT_MAX, it is 2^22. In case of CRAC_MIN_PID is greater than 2^22, we'd have a possible infinite loop while spinning the last PID.

Reply:
I wouldn't test what's in `/proc/sys/kernel/pid_max` explicitly, but if PID #N+1 is lower than PID #N we have probably overflowed and that's what we should detect, and fail gracefully.

Reply:
Got it, in case user has provided some big value. For spin_last_pid, yes, it makes a lot of sense to build some safety, like Radim's suggestion. Or to limit number of attempts e.g. by the value itself (it should not take more than min_pid attempts to get pid>min_pid).

Reply:
Should Java fails if PID cannot be moved to a desired PID value? Or just warn and go on?

Reply:
If the value was explicitly set, I think it would be better to fail. When it's trying to get to PID 128 by default I think it is sufficient to warn the user **and** tell him that he could switch off the warning setting `-XX:CRMinPid=1`.

Reply:
I did some experiments with PID spinning and a desired PID value that exceeds max_pid. It takes too long to spin PID until PID overflows. In case of a wrong value set by an user, this may seem like java hangs, so the user cannot wait so long to see the error message. This is also true for a valid desired PID value which is pretty big, e.g. 2_000_000.

We could remove `waitpid()` call to speed up PID spinning, but by removing this, we can easily reach container's resource limits (I tested this), so we cannot remove `waitpid` easily.

To avoid reading from `kernel/pid_max` and to avoid hanging on PID spinning, we could introduce max number of spin tries, say 10_000. If we reach this limit while spinning PIDs, we'd stop spinning and continue run Java with the currently reached PID. It actually seems doubtful that users want to move PID to 2M starting with PID=1 or 8 in a container. If users have some processes running in their container, on checkpoint they'd adjust desirable PID value in accordance with the state of the container, limited by a max try count we introducing. This solution seems portable for POSIX-like platforms.

Or, since things're becoming so complicated, it'd be easier to read max_pid, only for Linux though.

Please note that I'm talking about PID spinning, i.e. a case when writing to `ns_last_pid` haven't worked for some reasons.

Reply:
I agree that setting min pid to 2M is unlikely to be practical. Spinning fixed number of times is simpler, but if anything like this happens it would make sense to print a warning message (e.g. We are cycling PIDs due to -XX:CRMinPID=200000) after certain elapsed time (3 seconds?) rather than just number of iterations. Then the user can decide to cancel. Please make sure the stream gets flushed (in case this is written to a file or so...).

Reply:
Implemented PID spinning up to 10000 times. The 10000 is hardcoded now, we'd introduce an option if needed in future.